src/librustc/diagnostics.rs

```diff
@@ -1431,6 +1431,21 @@ fn make_child<'elve>(x: &mut &'elve isize, y: &mut &'elve isize) {
 ```
 "##,
 
+E0317: r##"
+An `if` expression without an `else` block is required to have the type `()`.
+This error occurs when the `if` block has a type other than `()`. For example:
+
+```compile_fail,E0317
+fn main() {
+    let x = 5;
+    let a = if x == 5 { 1 };
```

Comment:
It seems to be dependent on the expected type in the expression's context:

```
fn foo() -> i32 {
    if true {
        5
    }
}
```

produces:

```
error[E0308]: if may be missing an else clause
 --> <anon>:2:5
  |
2 |     if true {
  |     ^ expected (), found i32
  |
```

However, change `i32` to `()` and it's:

```
error[E0308]: mismatched types
 --> <anon>:3:9
  |
3 |         5
  |         ^ expected (), found integral variable
  |
```